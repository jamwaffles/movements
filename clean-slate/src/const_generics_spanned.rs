use crate::parse_modal_spans::Span;
use nom::{
    character::complete::{anychar, digit1, space0},
    combinator::{map_opt, verify},
    number::complete::float,
    sequence::{preceded, separated_pair},
    IResult, ParseTo,
};

pub fn recognise_word<const C: char, const N: u8>(i: Span) -> IResult<Span, ()> {
    let (i, _letter) = verify(anychar, |c| c.eq_ignore_ascii_case(&C))(i)?;

    let (i, _number): (_, u8) = verify(
        preceded(space0, map_opt(digit1, |d: Span| d.parse_to())),
        |n| *n == N,
    )(i)?;

    Ok((i, ()))
}

pub fn recognise_word_decimal<const C: char, const N: u8, const M: u8>(
    i: Span,
) -> IResult<Span, ()> {
    let (i, _) = separated_pair(
        recognise_word::<C, N>,
        space0,
        nom::character::complete::char('.'),
    )(i)?;

    let (i, _decimal): (_, u8) = preceded(
        space0,
        verify(map_opt(digit1, |d: Span| d.parse_to()), |n| *n == M),
    )(i)?;

    Ok((i, ()))
}

pub fn literal<const C: char>(i: Span) -> IResult<Span, f32> {
    let (i, _letter) = verify(anychar, |c| c.eq_ignore_ascii_case(&C))(i)?;

    preceded(space0, float)(i)
}

#[cfg(test)]
mod tests {
    use super::*;
    use nom::combinator::all_consuming;

    #[test]
    fn const_generics_int_word() {
        assert!(recognise_word::<'G', 0>("G0".into()).is_ok());
        assert!(recognise_word::<'G', 0>("G00".into()).is_ok());
        assert!(recognise_word::<'G', 0>("g00".into()).is_ok());
        assert!(recognise_word::<'G', 0>("g0".into()).is_ok());
        assert!(recognise_word::<'G', 0>("g01".into()).is_err());

        assert!(recognise_word::<'G', 4>("G04".into()).is_ok());
        assert!(recognise_word::<'G', 4>("G4".into()).is_ok());
    }

    #[test]
    fn const_generics_decimal_word() {
        assert!(recognise_word_decimal::<'G', 17, 1>("G17.1".into()).is_ok());
        assert!(recognise_word_decimal::<'G', 17, 1>("g17.1".into()).is_ok());
        assert!(recognise_word_decimal::<'G', 17, 1>("g17 . 1".into()).is_ok());
        assert!(recognise_word_decimal::<'G', 17, 1>("g   17   .    1".into()).is_ok());

        assert!(recognise_word_decimal::<'G', 17, 1>("M17.1".into()).is_err());
        assert!(recognise_word_decimal::<'G', 17, 1>("M17".into()).is_err());
        assert!(recognise_word_decimal::<'G', 17, 1>("M17.".into()).is_err());
        assert!(recognise_word_decimal::<'G', 17, 1>("M17.2".into()).is_err());
        assert!(recognise_word_decimal::<'G', 17, 1>("M17.1".into()).is_err());
    }

    #[test]
    fn literals() {
        assert_eq!(
            literal::<'P'>("p 0.005".into()),
            Ok((unsafe { Span::new_from_raw_offset(7, 1, "", ()) }, 0.005f32))
        );

        // Decimals with spaces in them are not supported.
        assert!(all_consuming(literal::<'P'>)("p 0 . 005".into()).is_err());
    }
}
