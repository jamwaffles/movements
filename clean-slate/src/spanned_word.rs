//! Parse a couple of modal groups and collate them into a [`Word`] enum.
//!
//! Also contains the [`Spanned`] struct to wrap an item with its position in the input.

use crate::const_generics_spanned::{literal, recognise_word};
use core::time::Duration;
use nom::{
    branch::alt,
    bytes::complete::take_until,
    character::complete::anychar,
    combinator::map,
    multi::many0,
    sequence::{delimited, preceded},
    IResult,
};
use nom::{character::complete::space0, sequence::separated_pair};
use nom_locate::position;
use nom_locate::LocatedSpan;

pub type Span<'a> = LocatedSpan<&'a str>;

/// A parsed word with its input position.
#[derive(Debug, PartialEq)]
pub struct Spanned<'a, T> {
    pub start: Span<'a>,
    pub end: Span<'a>,
    pub item: T,
}

pub fn spanned<'a, T>(
    mut inner: impl FnMut(Span<'a>) -> IResult<Span<'a>, T>,
) -> impl FnMut(Span<'a>) -> IResult<Span, Spanned<'a, T>> {
    move |i: Span<'a>| {
        let (i, start) = position(i)?;

        let (i, item) = inner(i)?;

        let (i, end) = position(i)?;

        Ok((i, Spanned { start, end, item }))
    }
}

#[derive(Debug)]
pub enum Word {
    /// Comment.
    Comment(Comment),

    // Modal groups.
    /// Group 0.
    NonModal(NonModal),

    /// Group 1.
    Motion(Motion),
}

// TODO: Trait?
impl Word {
    pub fn parse(i: Span) -> IResult<Span, Spanned<Self>> {
        spanned(alt((
            map(Comment::parse, Self::Comment),
            map(Motion::parse, Self::Motion),
            map(NonModal::parse, Self::NonModal),
        )))(i)
    }
}

#[derive(Debug)]
pub enum CommentKind {
    Block,
    Inline,
}

#[derive(Debug)]
pub struct Comment {
    pub kind: CommentKind,
    pub comment: String,
}

// TODO: Trait?
impl Comment {
    pub fn parse(i: Span) -> IResult<Span, Self> {
        alt((
            map(
                delimited(
                    nom::character::complete::char('('),
                    take_until(")"),
                    nom::character::complete::char(')'),
                ),
                |comment: Span| Comment {
                    kind: CommentKind::Inline,
                    comment: comment.fragment().to_string(),
                },
            ),
            map(
                preceded(nom::character::complete::char(';'), many0(anychar)),
                |comment: Vec<char>| {
                    let comment: String = comment.into_iter().collect();

                    Comment {
                        kind: CommentKind::Block,
                        comment,
                    }
                },
            ),
        ))(i)
    }
}

/// Group 1: Motion.
#[derive(Debug, PartialEq)]
pub enum Motion {
    // `G0`.
    Rapid,
}

// TODO: Trait?
impl Motion {
    pub fn parse(i: Span) -> IResult<Span, Self> {
        map(recognise_word::<'G', 0>, |_| Self::Rapid)(i)
    }
}

/// Group 0: Non-modal.
#[derive(Debug, PartialEq)]
pub enum NonModal {
    // `G4 Pn`.
    Dwell { duration: Duration },
}

// TODO: Trait?
impl NonModal {
    pub fn parse<'a>(i: Span) -> IResult<Span, Self> {
        map(
            separated_pair(recognise_word::<'G', 4>, space0, literal::<'P'>),
            |(_, duration)| Self::Dwell {
                duration: Duration::from_secs_f32(duration),
            },
        )(i)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn snapshot_motion() {
        insta::assert_debug_snapshot!(Motion::parse("G0".into()));
    }

    #[test]
    fn snapshot_non_modal() {
        insta::assert_debug_snapshot!(NonModal::parse("G4 P0.1".into()));
    }

    #[test]
    fn position() {
        insta::assert_debug_snapshot!(Word::parse("G4 P0.1".into()));
    }

    #[test]
    fn block_comment() {
        insta::assert_debug_snapshot!(Comment::parse("; absolute magic".into()));
    }

    #[test]
    fn inline_comment() {
        insta::assert_debug_snapshot!(Comment::parse("(inline comment) G0".into()));
    }
}
