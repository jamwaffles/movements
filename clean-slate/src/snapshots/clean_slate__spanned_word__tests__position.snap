---
source: clean-slate/src/spanned_word.rs
expression: "Word::parse(\"G4 P0.1\".into())"

---
Ok(
    (
        LocatedSpan {
            offset: 7,
            line: 1,
            fragment: "",
            extra: (),
        },
        NonModal(
            Dwell {
                duration: 100ms,
            },
        ),
    ),
)
